pkg/customunit/client.go

```diff
@@ -0,0 +1,145 @@
+package customunit
+
+import (
+	"context"
+	"time"
+
+	"github.com/Axway/agent-sdk/pkg/agent/cache"
+	cu "github.com/Axway/agent-sdk/pkg/amplify/agent/customunits"
+	"github.com/Axway/agent-sdk/pkg/util/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+type customUnitClient struct {
+	logger    log.FieldLogger
+	quotaInfo *cu.QuotaInfo
+	dialOpts  []grpc.DialOption
+	cOpts     []grpc.CallOption
+	url       string
+	conn      *grpc.ClientConn
+	isRunning bool
+	cache     cache.Manager
+	stopChan  chan struct{}
+	delay     time.Duration
+}
+
+type CustomUnitOption func(*customUnitClient)
+
+type CustomUnitClientFactory func(...CustomUnitOption) (*customUnitClient, error)
+
+func NewCustomUnitClientFactory(url string, agentCache cache.Manager, quotaInfo *cu.QuotaInfo) CustomUnitClientFactory {
+	return func(opts ...CustomUnitOption) (*customUnitClient, error) {
+		c := &customUnitClient{
+			quotaInfo: quotaInfo,
+			url:       url,
+			dialOpts: []grpc.DialOption{
+				grpc.WithTransportCredentials(insecure.NewCredentials()),
+			},
+			cache:     agentCache,
+			logger:    log.NewFieldLogger().WithPackage("customunit").WithComponent("client").WithField("metricServer", url),
+			stopChan:  make(chan struct{}),
+			isRunning: true,
+			delay:     30 * time.Second,
+		}
+
+		for _, o := range opts {
+			o(c)
+		}
+
+		return c, nil
+	}
+}
+
+func WithGRPCDialOption(opt grpc.DialOption) CustomUnitOption {
+	return func(c *customUnitClient) {
+		c.dialOpts = append(c.dialOpts, opt)
+	}
+}
+
+func (c *customUnitClient) createConnection() error {
+	conn, err := grpc.DialContext(context.Background(), c.url, c.dialOpts...)
```

Comment:
Maybe we do not need the grpc.DialContext call since we are not using the context for these

Reply:
DialContext is deprecated and replaced with grpc.NewClient()